keps/sig-storage/3294-provision-volumes-from-cross-namespace-snapshots/README.md

```diff
@@ -0,0 +1,1132 @@
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+Existing PV provisioning also fails.
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+- Implement transfer feature for `VolumeSnapshot`:
+  - Pros:
+    - Can have more control over the transferred `VolumeSnapshot`, like modifying and deleting
+    - Can potentially be used to directly clone the `VolumeSnapshot` to other namespaces for [backup use case](https://github.com/kubernetes/enhancements/pull/2849#issuecomment-957693334)
+  - Cons:
+    - Need to handle [race conditions](https://github.com/kubernetes/enhancements/pull/2849#discussion_r682057570)
+    - Need to consider [referenced Secrets](https://github.com/kubernetes/enhancements/pull/2849#discussion_r692459041) from snapshot after transferred
+
+- Change `PersistentVolumeClaimSpec` in core API to take namespace as a data source, instead of introducing a new `VolumeSnapshotLink` API:
+  - Pros:
+    - Can provide a simple API, which will provide a better UX
```

Comment:
If we use the similar approach to AnyVolumeDatasource feature, it will be like as follows:

 1. Add a new field like `DataSourceRef2`, which has all fields in `TypedLocalObjectReference` and namespace field, to [PersistentVolumeClaimSpec](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L422-L478)

     Any suggestions on better name?

 2. Add conversion codes among `Datasource`, `DataSourceRef`, and `DataSourceRef2` like [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/persistentvolumeclaim/strategy.go#L100-L119) and [here](https://github.com/kubernetes/kubernetes/blob/b4eec3c2ad8d0645a10b4e911738627b7ecb2809/pkg/api/persistentvolumeclaim/util.go#L119-L131).

3. Use `DataSourceRef2` field instead of `VolumeSnapshotLink` object.

 Then, should we avoid adding codes using `VolumeSnapshotLink` API as alpha, if we know that we eventually take core API change approach?

Reply:
Make this option as a proposal and moved the `SnapshotLink` approach to alternative.